### ISSUE TITLE
"ValueError: list.remove(x): x not in list" in eager mode for decorated keras method

### ISSUE THREAD

**Issue by simonm3** [tensorflow/tensorflow#23774]
The below runs without the decorator but when the decorator is added then it fails with the above error.

tf.__version__, keras.__version__ = ('1.12.0', '2.1.6-tf')
ubuntu 16.04
```
import tensorflow.keras.layers as KL
from functools import wraps

def decorator(f):
    @wraps(f)
    def wrapper(*args, **kwargs):
        return f(*args, **kwargs)
    return wrapper

class MyLayer(KL.Layer):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    @decorator
    def call(self, inputs):
        pass
a=MyLayer()
```

**Comment by simonm3** [comment]
I discovered a solution. Just pip install decorator; then decorate my decorator with @decorator.decorator. This keeps the original argument names for inspect to find.

**Comment by simonm3** [comment]
On second thoughts this should be fixed in tensorflow as workarounds are fiddly. Typically a decorator would take *args, **kwargs if it could be used on a function or class method. In function_utils.py line 58 it has  args.remove('self'). Changing that to args.pop(0) would I think resolve.

**Comment by facaiy** [comment]
I'm afraid `args.pop(0)` cannot solve it, could you explain why to use decorator for `__call__` method?

**Comment by simonm3** [comment]
I have a decorator that saves the parameters and results of decorated
functions. This is useful to compare two versions. For example I have
working versionA and decorate it so it saves baseline inputs and results
(in eager mode). Then I use those inputs on versionB and compare the
results. This allows me to simplify or make code faster but to test it
still produces same results; or to port between keras/pytorch and compare
the two versions.

Why wouldn't args.pop(0) solve it? It is only executed for a bound
function.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
This is fixed in latest tf-nightly version 1.15.0-dev20190725. Thanks!
https://github.com/tensorflow/tensorflow/blob/4ad169057cf0adb51767914e84c300c95ce748d5/tensorflow/python/util/function_utils.py#L62
